[SPARK-11327] [MESOS] Dispatcher does not respect all args from the Submit request

jayv opened the pull request:
Supersedes https://github.com/apache/spark/pull/9752

jayv committed:
Messos scheduler does not respect all args from the Submit request

AmplabJenkins commented:
Can one of the admins verify this patch?

tnachen commented:
ok to test

tnachen commented:
At first I was wondering if we just copy all scheduler properties will it clash with the ones we used above (i.e --total-executor-cores), but looks like the command line flags takes precedence in SparkSubmit so we should be ok. How have you tested this?

SparkQA commented:
**[Test build #47989 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47989/consoleFull)** for PR 10370 at commit [`cdde93d`](https://github.com/apache/spark/commit/cdde93d45401a2f7a97a46b25387c297dcf21d2f).

SparkQA commented:
**[Test build #47989 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47989/consoleFull)** for PR 10370 at commit [`cdde93d`](https://github.com/apache/spark/commit/cdde93d45401a2f7a97a46b25387c297dcf21d2f).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47989/
Test FAILed.

jayv commented:
@tnachen I've tested that patch against branch-1.5 and I'm now able  to specify options such as ```--conf spark.executor.userClassPathFirst=true --conf spark.driver.userClassPathFirst=true``` on spark-submit for a cluster deploy via mesos.

tnachen commented:
@jayv Looks like you need to fix the line length > 100 style rule.

dragos commented on core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala:
```diff
@@ -440,6 +444,9 @@ private[spark] class MesosClusterScheduler(
     }
+    desc.schedulerProperties
+      .filter { case (key, _) => !replicatedOptionsBlacklist.contains(key) }
+      .foreach { case (key, value) => options ++= Seq("--conf", Seq(key, "=\"", value, "\"").mkString("")) }
```
Are the quotes inside `--conf key="value"` valid? I know I tried `--conf spark.mesos.executor.cores="/var/spark"` and that failed (the actual value of `spark.mesos.executor.home` contained the quotes.

dragos replied:
..one more thing: better use string interpolation instead of `Seq.mkString`: s"$key=$value"`.

jayv replied:
Yes, the quotes work for me.

In order to pass multiple system properties via eg `spark.driver.extraJavaOptions`, I passed the arg in quotes on the shell to spark-submit, but the value itself doesn't have quotes around it and ends up broken on the commandline for the driver, unless I add escaped quotes inside the string.

So this fails when executing the driver, but spark-submit accepts the command:
```
bin/spark-submit [...] spark.driver.extraJavaOptions="-Dcom.sun.jersey.server.impl.cdi.lookupExtensionInBeanManager=true -Dorg.jboss.weld.bootstrap.concurrentDeployment=false" [...]
```
This doesn't look very intuitive to me, but it does work without wrapping values in quotes:
```
bin/spark-submit [...] spark.driver.extraJavaOptions="\"-Dcom.sun.jersey.server.impl.cdi.lookupExtensionInBeanManager=true -Dorg.jboss.weld.bootstrap.concurrentDeployment=false\"" [...]
```

String interpolation doesn't support double quotes, but I've since learned about triple-double-quoted string interpolation, that why I initially went with `Seq.mkString`.

dragos commented:
thanks @jayv, I left one note, otherwise this looks good.

jayv committed:
Reduce line length, use interpolation with wrapped quotes

SparkQA commented:
**[Test build #48229 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48229/consoleFull)** for PR 10370 at commit [`8f0e1f2`](https://github.com/apache/spark/commit/8f0e1f2ad5e054bdc13423958ebc41fc9bbe6e9e).

SparkQA commented:
**[Test build #48229 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48229/consoleFull)** for PR 10370 at commit [`8f0e1f2`](https://github.com/apache/spark/commit/8f0e1f2ad5e054bdc13423958ebc41fc9bbe6e9e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `abstract class ImperativeAggregate extends AggregateFunction with CodegenFallback `\n  * `case class UnresolvedWindowExpression(`\n  * `case class WindowExpression(`\n  * `case class Lead(input: Expression, offset: Expression, default: Expression)`\n  * `case class Lag(input: Expression, offset: Expression, default: Expression)`\n  * `abstract class AggregateWindowFunction extends DeclarativeAggregate with WindowFunction `\n  * `abstract class RowNumberLike extends AggregateWindowFunction `\n  * `trait SizeBasedWindowFunction extends AggregateWindowFunction `\n  * `case class RowNumber() extends RowNumberLike `\n  * `case class CumeDist() extends RowNumberLike with SizeBasedWindowFunction `\n  * `case class NTile(buckets: Expression) extends RowNumberLike with SizeBasedWindowFunction `\n  * `abstract class RankLike extends AggregateWindowFunction `\n  * `case class Rank(children: Seq[Expression]) extends RankLike `\n  * `case class DenseRank(children: Seq[Expression]) extends RankLike `\n  * `case class PercentRank(children: Seq[Expression]) extends RankLike with SizeBasedWindowFunction `\n

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/48229/
Test PASSed.

dragos commented:
This PR might help with #8872 testing

dragos commented:
Ok, it seems you need to filter out a few more things, particularly `spark.submit.deployMode="cluster"`. I see errors[1] in the driver `stderr`, since the driver tries to launch *again* in cluster mode, on one of the machines in the cluster. If this worked for you, I'd like to know the exact sequence of steps (I'd really like to get this fixed, though). Here are mine:

```
$ sbin/start-mesos-dispatcher.sh --master mesos://lausanne1.local:5050
$ bin/spark-submit --deploy-mode cluster --master mesos://sagitarius.local:7077 --conf spark.mesos.coarse=true --conf spark.executor.uri="ftp://sagitarius.local/ftp/spark-1.6.0-SNAPSHOT-bin-pr10370.tgz" --class org.apache.spark.examples.SparkPi ftp://sagitarius.local/ftp/spark-examples-1.6.0-SNAPSHOT-hadoop2.2.0.jar
```
`sagitarius` is my machine, `lausanne1` is the Mesos master in my local network.

[1]
```
16/01/18 16:56:31 INFO RestSubmissionClient: Submitting a request to launch an application in mesos://lausanne1.local:5050.
Exception in thread "main" org.apache.spark.deploy.rest.SubmitRestProtocolException: Malformed response received from server
	at org.apache.spark.deploy.rest.RestSubmissionClient.readResponse(RestSubmissionClient.scala:264)
	at org.apache.spark.deploy.rest.RestSubmissionClient.org$apache$spark$deploy$rest$RestSubmissionClient$$postJson(RestSubmissionClient.scala:222)
	at org.apache.spark.deploy.rest.RestSubmissionClient$$anonfun$createSubmission$3.apply(RestSubmissionClient.scala:87)
	at org.apache.spark.deploy.rest.RestSubmissionClient$$anonfun$createSubmission$3.apply(RestSubmissionClient.scala:83)
	at scala.collection.TraversableLike$WithFilter$$anonfun$foreach$1.apply(TraversableLike.scala:772)
	at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
	at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:108)
	at scala.collection.TraversableLike$WithFilter.foreach(TraversableLike.scala:771)
	at org.apache.spark.deploy.rest.RestSubmissionClient.createSubmission(RestSubmissionClient.scala:83)
	at org.apache.spark.deploy.rest.RestSubmissionClient$.run(RestSubmissionClient.scala:411)
	at org.apache.spark.deploy.rest.RestSubmissionClient$.main(RestSubmissionClient.scala:424)
	at org.apache.spark.deploy.rest.RestSubmissionClient.main(RestSubmissionClient.scala)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.apache.spark.deploy.SparkSubmit$.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:731)
	at org.apache.spark.deploy.SparkSubmit$.doRunMain$1(SparkSubmit.scala:181)
	at org.apache.spark.deploy.SparkSubmit$.submit(SparkSubmit.scala:206)
	at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:121)
	at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)
Caused by: org.apache.spark.deploy.rest.SubmitRestProtocolException: Server returned empty body
	at org.apache.spark.deploy.rest.RestSubmissionClient$$anonfun$1.apply(RestSubmissionClient.scala:241)
	at org.apache.spark.deploy.rest.RestSubmissionClient$$anonfun$1.apply(RestSubmissionClient.scala:232)
	at scala.concurrent.impl.Future$PromiseCompletingRunnable.liftedTree1$1(Future.scala:24)
	at scala.concurrent.impl.Future$PromiseCompletingRunnable.run(Future.scala:24)
	at scala.concurrent.impl.ExecutionContextImpl$$anon$3.exec(ExecutionContextImpl.scala:107)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

dragos committed:
Don’t pass deployMode and spark.master when launching a driver.

- the launched driver should be in client mode, but at submission time this setting is set to ‘cluster’.
- the `spark.master` setting is set to the dispatcher address (usually on port 7077). Passing it has no effect,
since we also pass `—-master` which takes precedence. But this is confusing so we should filter it out.

dragos commented:
I opened jayv/spark#2 with two more filters: `spark.submit.deployMode` and `spark.master`. I think it's ok to remove `[WIP]` from the title, and if you agree with my changes, and passes your tests (it passed mine), this is an LGTM

jayv commented:
@dragos I don't remember seeing these errors on the 1.5.x branch I tested against,but I may have missed them since I was focussing on the missing args. This looks totally reasonable to me.

SparkQA commented:
**[Test build #49706 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49706/consoleFull)** for PR 10370 at commit [`4b2b8b1`](https://github.com/apache/spark/commit/4b2b8b178820288610a91c81dd5b60864f1c27d7).

SparkQA commented:
**[Test build #49706 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49706/consoleFull)** for PR 10370 at commit [`4b2b8b1`](https://github.com/apache/spark/commit/4b2b8b178820288610a91c81dd5b60864f1c27d7).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49706/
Test FAILed.

dragos commented:
The failed test is in Spark Streaming, no doubt it's a flaky one.

LGTM. @andrewor14 can you please have a look?

mgummelt commented on core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala:
```diff
@@ -440,6 +446,9 @@ private[spark] class MesosClusterScheduler(
     }
+    desc.schedulerProperties
+      .filter { case (key, _) => !replicatedOptionsBlacklist.contains(key) }
+      .foreach { case (key, value) => options ++= Seq("--conf", s"""$key="$value"""") }
```
Are lines 437-443 now redundant?  If setting --total-executor-cores is identical to --conf spark.cores.max, then we don't need to set both.

andrewor14 replied:
does this escape quotes, double quotes, whitespaces and backslashes correctly?

dragos replied:
That's a good point, `CommandInfo` is using `/bin/sh` to launch the command. :confused: 

Spaces should be ok, everything else won't be correctly escaped. Skimming through Spark properties I think the only ones that could pose problems are `spark.authenticate.secret` and the other passwords (SSL, etc.). Still, this needs a solution.

tnachen replied:
Mesos should also handle spaces if you provide the arguments as args in the CommandInfo even with spaces and other characters.

dragos replied:
I guess Tim refers to [this line](https://github.com/apache/mesos/blob/master/include/mesos/mesos.proto#L387), so we'd need to set `shell` to `false`. That's probably the simplest way.

jayv replied:
Ah OK, I can check that, I wrote a bunch of regexes and tests today to solve escaping for most Unix shells. Was about to push a PR early morning.

dragos replied:
I think the Mesos way is more robust and requires less code. Hopefully :)

jayv replied:
The meat of it is only 5 lines :) but let's see if we can make it work without side-effects and for my use-case with args with spaces.

andrewor14 commented:
retest this please

SparkQA commented:
**[Test build #49806 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49806/consoleFull)** for PR 10370 at commit [`4b2b8b1`](https://github.com/apache/spark/commit/4b2b8b178820288610a91c81dd5b60864f1c27d7).

SparkQA commented:
**[Test build #49806 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49806/consoleFull)** for PR 10370 at commit [`4b2b8b1`](https://github.com/apache/spark/commit/4b2b8b178820288610a91c81dd5b60864f1c27d7).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49806/
Test PASSed.

dragos commented:
@jayv do you have time to work on this, or should I?

jayv commented:
Today I will test the driver issue you mentioned, we also seem to have a rogue driver spinning up on workers, and work on better escaping. Isn't there code for this in spark-submit or something already?

jayv committed:
Escape args for Unix-like shells

(cherry picked from commit 8023da309df2e5440e244e12d449cb8107d15b00)

andrewor14 commented:
@jayv the closest thing I can think of is `Utils.splitCommandString`, which does the unescaping. Did you end up getting Mesos to escape things properly? E.g. does something like `spark.executor.extraJavaOptions` work?

andrewor14 commented:
@dragos is the only thing that's left before this patch can be merged the escaping thing?

jayv commented:
Got an implementation that escapes all the things under linux shells, but had to suspend work on this, so was unable to test mesos' native escaping. I'll try to get back to it this week.

dragos commented:
As far as I'm concerned, that's the only thing (I'll have to test again on a real Mesos cluster).

dragos commented:
@jayv did you have the chance to look at this again?

dragos commented:
I pushed [this commit](https://github.com/dragos/spark/commit/a6f9df1) with the general idea, but I didn't get to test it much. I'll come back to it tomorrow.

jayv commented:
@dragos Unfortunately not, I'll come back to this issue when I've done some more testing.

dragos commented:
My commit is not working, BTW, I'll come back to this (the executable needs to be a complete path to `spark-class`, right now it's a bash-ism: `cd spark-1.*; ./bin/spark-class`.

tnachen commented:
@dragos Ah yes if you need bash completion than we have to use shell = true.

dragos commented:
Shouldn't be a problem, I just missed it in my first try. I think it's not too hard to make it work with `shell=false`, I just need to spend a bit more time on this. I'll probably open a new PR then.

dragos commented:
OK, I retract that. We need to use `shell=true`, so @jayv if you have that character escape version, let's go that way.

The reason why it won't work is `extraLibraryPath`: we need to prepend library path variables, like `LD_LIBRARY_PATH="path1:path2:$LD_LIBRARY_PATH"`. We could add this binding in the environment that we're building, but without a shell we can't refer to the old value of `LD_LIBRARY_PATH` (note that this has to stay an unevaluated reference to `LD_LIBRARY_PATH` because we're preparing the `CommandInfo` structure in the dispatcher, which is NOT the machine where the command will be executed).

dragos commented:
@jayv please let me know if you will look into this one. If you're busy, I'm happy to take over, and will happily start with your escaping algorithm ;-)

jayv commented:
Sorry for the delay @dragos very busy indeed, maybe this can help https://github.com/jayv/spark/commit/8023da309df2e5440e244e12d449cb8107d15b00

tnachen commented:
@jayv are you planning to update this PR with that commit?

dragos commented:
I can pick this up tomorrow.

dragos commented:
It works, but I found a couple of corner cases. I guess you need to escape `\` as well. For instance, a string ending in `\`, or a string like `\"?` won't be quoted correctly. Otherwise it looks good and the manual tests are passing.

jayv commented:
Nice, will you pick this up @dragos or waiting for me?  - can't commit any time this week

dragos committed:
Escape `\` in strings.

dragos commented:
Yeap, I just opened a PR on your repo. If you merge it it should be reflected here.

SparkQA commented:
**[Test build #53977 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53977/consoleFull)** for PR 10370 at commit [`11b070d`](https://github.com/apache/spark/commit/11b070df733ec45432214691d7d83968c3d0e180).

SparkQA commented:
**[Test build #53977 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53977/consoleFull)** for PR 10370 at commit [`11b070d`](https://github.com/apache/spark/commit/11b070df733ec45432214691d7d83968c3d0e180).
 * This patch **fails R style tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53977/
Test FAILed.

dragos commented:
@jayv can you please rebase?

@andrewor14 the last issue (escaping characters for the shell command) has been fixed. Please take a look.

jayv commented:
@dragos rebased!

SparkQA commented:
**[Test build #54541 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54541/consoleFull)** for PR 10370 at commit [`82cd03d`](https://github.com/apache/spark/commit/82cd03dfc794c960f48d8e2e91a239a885196161).

SparkQA commented:
**[Test build #54541 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54541/consoleFull)** for PR 10370 at commit [`82cd03d`](https://github.com/apache/spark/commit/82cd03dfc794c960f48d8e2e91a239a885196161).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54541/
Test FAILed.

andrewor14 commented:
retest this please

SparkQA commented:
**[Test build #54556 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54556/consoleFull)** for PR 10370 at commit [`82cd03d`](https://github.com/apache/spark/commit/82cd03dfc794c960f48d8e2e91a239a885196161).

andrewor14 commented:
Latest changes look pretty reasonable. Funny how mesos cluster mode is basically unusable before. :)

SparkQA commented:
**[Test build #54556 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54556/consoleFull)** for PR 10370 at commit [`82cd03d`](https://github.com/apache/spark/commit/82cd03dfc794c960f48d8e2e91a239a885196161).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54556/
Test PASSed.

dragos commented:
LGTM! @andrewor14 please have a look

andrewor14 commented:
Merged into master. Thanks for your patience and hard work everyone.

andrewor14 commented:
@dragos @tnachen can you remind me on the status of Mesos cluster mode in 1.6? It was essentially unusable because of the external shuffle service right? Did we ever resolve that? (where is the patch?)

mgummelt commented:
Mesos Cluster Mode works fine.  It's dynamic allocation that was broken.  I'm pretty sure this is the PR: https://github.com/apache/spark/pull/11272

@dragos @skonto can you verify that the Spark/Mesos dynamic allocation integration tests on master are passing?

andrewor14 commented:
Ah thanks, I forgot I merged that.

andrewor14 commented:
Then it seems that this patch is worth backporting into 1.6 since cluster mode does work there.

@dragos @mgummelt @jayv Could one of you submit a patch for 1.6? I think the changes here are pretty isolated so it should be safe.
